### PR TITLE
Large document changes

### PR DESCRIPTION
**Pull request by thebsdbox**
This adds a large amount of changes for the upcoming `0.5` release.

### CODE REVIEW

File: content/_index.md [hunk 1]
```diff
@@ -17,6 +17,24 @@ kube-vip provides Kubernetes clusters with a virtual IP and load balancer for bo
 
 [![Build and publish main image regularly](https://github.com/kube-vip/kube-vip/actions/workflows/main.yaml/badge.svg)](https://github.com/kube-vip/kube-vip/actions/workflows/main.yaml)
 
+## Modes 
+
+### ARP
+
+**ARP** is a layer 2 protocol that is used to inform the network of the location of a new address. When a new IP address is configured to a device, there needs to be a mechanism to inform the network of which piece of hardware is hosting this new address. ARP is the technology that is used to ensure that a network understands the link between the hardware address (MAC) and the logical address (IP).
```
Comment (MikeZappa87) [inline content/_index.md:24]: Should it mention Gratuitous Arp? This definition comes off a little misleading.

File: README.md [hunk 2]
```diff
@@ -5,3 +5,9 @@ This guide covers the necessary bits. As the project evolves, it will only becom
 ## Updating
 
 This site uses the Compose Hugo theme loaded as a Hugo module. Updating the theme from the base site directory: `hugo mod get -u ./...`
+
+To run a live server as you're working on doc updates, please use the following command:
+
+```
+hugo server --bind 0.0.0.0 -b `hostname -f`
```
Comment (chipzoller) [inline README.md:12]: You really don't need anything past `hugo server` unless you specifically need to bind to all interfaces (useful when allowing another device to access the live server instance).

File: content/_index.md [hunk 3]
```diff
@@ -17,6 +17,24 @@ kube-vip provides Kubernetes clusters with a virtual IP and load balancer for bo
...
+
+### BGP
+
+**BGP** is a mechanism so that networks that rely on routing (layer 3) can ensure that new addresses are advertised to the routing infrastructure. When this information has been updated it transparently means that the router will automatically forward traffic to the correct devices.
+
+### Routing Table
+
+The **Routing Table** mode is to allow additional routing technologies such as ECMP etc. to be configuraed so that traffic can be send to a range of nodes (such as your Kubernetes nodes), and kube-vip will manage the addition/deletion of addresses to the routing tables of these nodes so that they can recieve the correct traffic.
```
Comment (chipzoller) [inline content/_index.md:32]: "configured"

File: content/_index.md [hunk 4]
```diff
@@ -17,6 +17,24 @@ kube-vip provides Kubernetes clusters with a virtual IP and load balancer for bo
...
+
+### WireGuard
+
+The [**Wireguard**](https://www.wireguard.com/) mode allows Kubernetes services to be advertised over the wireguard interface (`wg0`), it's main use-case is so that distributed services across multiple clusters can centralise all their advised services on a central network controlled by wireguard.
```
Comment (chipzoller) [inline content/_index.md:36]: Run-on sentence.
"Its main use case is..."

File: content/docs/about/productionusage.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+title: "Production Usage"
+weight: 14
+description: >
+  kube-vip usage
+---
+
+
+## kube-vip Maturity
+
+The kube-vip project is deployed and used in a number of production environments and is included/bundled with a number of production Kubernetes based platforms. Whilst kube-vip could be thought of as pre-release, it has happily found it's home in places where it has ran production workloads. I cursary glance at time of writing this (mid-July 2022) also reveals that downloads combined from Docker Hub and GitHub container registry amount to around 19,000,000 kube-vip container pulls.
```
Comment (chipzoller) [inline content/docs/about/productionusage.md:11]: First sentence: "...it has happily found its home in places where it runs production workloads."
Second sentence: "A cursory glance at this time of writing (July 2022) also..."

File: content/docs/modes/_index.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+---
+title: "Modes"
+weight: 20
+description: >
+  kube-vip Modes
+---
+
+There are various technologies availble to kube-vip that are all used in largely the same manner, to expose an IP address to a network. These are either to update the local network that an IP address is on a particular host to updating networking equipment specifically that traffic should be sent to node(s) in order to reach a specific address.
```
Comment (chipzoller) [inline content/docs/modes/_index.md:8]: "available"
Second sentence is confusing to read. Suggest rewording.

File: content/docs/modes/arp.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
+---
+title: "ARP"
+weight: 21
+description: >
+  kube-vip ARP mode
+---
+
```
Comment (chipzoller) [inline content/docs/modes/arp.md:7]: Clarify in your parenthetical comment that this reference is only in context to kube-vip. ARP is a protocol which isn't really called layer 2 (an OSI model reference).

Second sentence avoid ending with prepositions:

"...so that a host knows to which physical or virtual NIC traffic should be sent."

File: content/docs/modes/arp.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ARP (sometimes called layer 2) is a simplistic protocol that is used to update the underlying network that in order to reach a certain IP address traffic should be sent to a specific piece of hardware. ARP will typically broadcast updates to the entire network that update the IP to Hardware (MAC) mapping, so that a host knows which physical or virtual network nic should be used to send traffic to.
+
+### Control plane
+
+When using ARP with the control plane feature, then a `leaderElection` will take place in order to determine the leader. This leader will then host the control place IP address and use ARP (through a gratuitous ARP broadcast) to update the network accordingly.
```
Comment (chipzoller) [inline content/docs/modes/arp.md:12]: Avoid using code formatting for concepts. The concept/process here is leader election. In the Services section, this is activated via `leaderElection`.

File: content/docs/modes/arp.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+- `leaderElection` for all services
+- `leaderElection` per service
+
+### `leaderElection` for all services
+
+In this mode (default) all kube-vip pods will elect a leader and this leader will be in charge of exposing all service addresses through ARP. This mode can become a bottleneck as traffic for **all** services will come to a single node.
+
+### `leaderElection` per service
+
+In this mode kube-vip will perform an election evertime a new Kubernetes service is created, this will allow service addresses to be spread across all nodes where a kube-vip pod is running.
```
Comment (chipzoller) [inline content/docs/modes/arp.md:27]: "everytime"

Also, run-on sentence.

File: content/docs/modes/routingtables.md [hunk 10]
```diff
@@ -0,0 +1,8 @@
+---
+title: "Routing Tables"
+weight: 23
+description: >
+  kube-vip Routing Table mode
+---
+
+The **Routing Table** mode is to allow additional routing technologies such as ECMP etc. to be configuraed so that traffic can be send to a range of nodes (such as your Kubernetes nodes), and kube-vip will manage the addition/deletion of addresses to the routing tables of these nodes so that they can recieve the correct traffic.
```
Comment (chipzoller) [inline content/docs/modes/routingtables.md:8]: "configured"
Sentence is unwieldy. Suggest rework.

File: content/docs/modes/wireguard.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+---
+title: "Wireguard"
+weight: 24
+description: >
+  kube-vip Wireguard mode
+---
+
+The [**Wireguard**](https://www.wireguard.com/) mode allows Kubernetes services to be advertised over the wireguard interface (`wg0`), it's main use-case is so that distributed services across multiple clusters can centralise all their advised services on a central network controlled by wireguard.
```
Comment (chipzoller) [inline content/docs/modes/wireguard.md:8]: See previous comment on this section.

File: content/docs/troubleshooting/_index.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
+---
+title: "Troubleshooting"
+weight: 90
+description: >
+  kube-vip Troubleshooting
+---
+
+## Wireguard 
+
+To debug most wireguard issues you'll need to examine the node that has the leadership or the wireguard servier itself.
```
Comment (chipzoller) [inline content/docs/troubleshooting/_index.md:10]: "server"

File: content/docs/troubleshooting/_index.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+peer: Zl4q4n6aLOJcLvNxzbsknpUccvJxlXg/e3isNjBF5Gk=
+  endpoint: 192.168.0.140:51820
+  allowed ips: 10.0.0.0/8
+  latest handshake: 1 minute, 21 seconds ago
+  transfer: 4.25 MiB received, 1.05 MiB sent
+```
+
+### An `allowed ips: (none)` configuration
+
+In the event an peer has no allowed ips, it usually means that two peers were created with the same ip range (which is illegal in Wireguard)
```
Comment (chipzoller) [inline content/docs/troubleshooting/_index.md:32]: "In the event a peer has no allowed IPs it usually means that two peers were created with the same IP range (which is illegal in Wireguard)."

Comment (chipzoller) [reply content/docs/troubleshooting/_index.md:32]: "Wireguard" is a proper noun and should be written consistently throughout this PR.

File: content/docs/usage/wireguard.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+title: "Wireguard"
+weight: 56
+description: >
+  kube-vip usage on wireguard.
+---
+
+## Pre-requisistes
+
+- This will require kube-vip starting as a daemonset as it will need to read existing data (`secrets`) from inside the cluster.
+- A Wireguard server that will act as the merging network point for all external clusters services
```
Comment (chipzoller) [inline content/docs/usage/wireguard.md:11]: "...external cluster services" ??

File: content/docs/usage/wireguard.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+- Distinct network ranges for each cluster (an overlap of a range will break one clusters configuration)
```
Comment (chipzoller) [inline content/docs/usage/wireguard.md:12]: "(an overlap of a range will break one cluster's configuration)"

File: content/docs/usage/wireguard.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+
+## Architecture
+
+On each cluster a secret is created that contains the private key and the address of the wireguard server, on the Wireguard server itself the clusters public key and network range is configured.
```
Comment (chipzoller) [inline content/docs/usage/wireguard.md:17]: Run-on sentence.

File: content/docs/usage/wireguard.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+                         │
+                         │
+┌─────────────────────┬──┘
+│                     │
+│     Cluster A       │
+│                     │
+└─────────────────────┘
+```
+
+A /30 or /29 is recommended for the network range, this can then match the configuration in the cloud-controller **or** services can be created with fixed addresses.
```
Comment (chipzoller) [inline content/docs/usage/wireguard.md:43]: Run-on sentence.

File: content/docs/usage/wireguard.md [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+The below command will generate a configuration for a **single** cluster:
+
+The `NETWORKRANGE` needs to be different for each cluster
+```
+SERVERIP=192.168.0.197
+NETWORKRANGE=10.0.0.0/30
+```
+
+This latter part of this command will auto-generate the required credentials and configure the server (the command has to be ran on the server), it will then print the secret to be created on the cluster itself.
```
Comment (chipzoller) [inline content/docs/usage/wireguard.md:55]: "(the command has to be run on the server)"
Run-on sentence.

File: content/docs/usage/wireguard.md [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+```
+PRIKEY=$(wg genkey)
+PUBKEY=$(echo $PRIKEY | wg pubkey)
+PEERKEY=$(sudo wg show wg0 public-key)
+sudo wg set wg0 peer $PUBKEY allowed-ips $NETWORKRANGE
+echo "kubectl create -n kube-system secret generic wireguard --from-literal=privateKey=$PRIKEY --from-literal=peerPublicKey=$PEERKEY --from-literal=peerEndpoint=$SERVERIP"
+
```
Comment (chipzoller) [inline content/docs/usage/wireguard.md:62]: Excess new line.
